### PR TITLE
Relate receiver invariantly in method probe for `Mode::Path`

### LINKED ISSUE #126227: inherent method selection uses subtyping even when using paths
```rust
struct B<T>(T);

impl B<fn(&'static ())> {
    fn method(self) {
        println!("hey");
    }
}

fn foo(x: B<for<'a> fn(&'a ())>) {
    x.method(); // ok, method call
    B::<for<'a> fn(&'a ())>::method(x); // ok, explicit path
}

fn main() {
   
}
```
I would like us to change method selection for associated items to use equality. It feels very wrong to completely ignore the user type annotation here:
```rust
struct B<T>(T);

impl B<fn(&'static ())> {
    fn method(self) {
        println!("hey");
    }
}

fn foo(y: B<fn(&'static ())>) {
    // looks like it requires a hr fn pointer,
    // but we never check that.
    B::<for<'a> fn(&'a ())>::method(y);
}

fn main() {}
```
cc #126128 @compiler-errors @oli-obk

### PR DESCRIPTION
**Pull request by compiler-errors**
Effectively reverts part of #126128
Fixes #126227

This PR changes method probing to use equality for fully path-based method lookup, and subtyping for receiver `.` method lookup.

r? lcnr

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/probe.rs [hunk 1]
```diff
@@ -1442,7 +1450,8 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
                     (xform_self_ty, xform_ret_ty) =
                         self.xform_self_ty(probe.item, impl_ty, impl_args);
                     xform_self_ty = ocx.normalize(cause, self.param_env, xform_self_ty);
-                    match ocx.sup(cause, self.param_env, xform_self_ty, self_ty) {
+                    match ocx.relate(cause, self.param_env, self.variance(), xform_self_ty, self_ty)
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/method/probe.rs:1453]: that's cursed... I feel like we should use `ty::Covariant` for method calls and actually flip `xform_self_ty` and `self_ty` here :blush:

File: tests/ui/coercion/coerce-issue-49593-box-never.rs [hunk 2]
```diff
@@ -16,7 +16,7 @@ fn foo(x: !) -> Box<dyn Error> {
     // Subtyping during method resolution will generate new inference vars and
     // subtype them. Thus fallback will not fall back to `!`, but `()` instead.
```
Comment (lcnr) [inline tests/ui/coercion/coerce-issue-49593-box-never.rs:17]: subtype :<

**Comment by lcnr** [comment]
should do a crater run for this

File: compiler/rustc_hir_typeck/src/method/probe.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/method/probe.rs:1453]: but then we change the "expected" and "found" 🤔 of the type errors I wonder if that affects diagnostics

**Comment by lcnr** [comment]
@craterbot check

File: compiler/rustc_hir_typeck/src/method/probe.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_hir_typeck/src/method/probe.rs:1453]: probably, the question is whether it makes them worse :grin:

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/method/probe.rs:1453]: Yeah I'll test it

**Comment by compiler-errors** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-129073/retry-regressed-list.txt p=1

File: tests/ui/coercion/coerce-issue-49593-box-never.rs [hunk 3]
```diff
@@ -14,9 +14,9 @@ fn raw_ptr_box<T>(t: T) -> *mut T {
 
 fn foo(x: !) -> Box<dyn Error> {
     // Subtyping during method resolution will generate new inference vars and
```
Comment (lcnr) [inline tests/ui/coercion/coerce-issue-49593-box-never.rs:16]: `// Method resolution will generate`...

**Comment by lcnr** [comment]
This PR changes path method lookup to use equality instead of subtyping, causing the following test to go from pass to fail. This does not cause any crater breakage and matches my expected behavior.
```rust
struct B<T>(T);

impl B<fn(&'static ())> {
    fn method(self) {
        println!("hey");
    }
}

fn foo(y: B<fn(&'static ())>) {
    B::<for<'a> fn(&'a ())>::method(y);
    //[with_this_change]~^ ERROR no function or associated item named `method` found
}
```
@compiler-errors you state that the existing behavior was introduced by #122317, however, the test case already compiled in previous versions: https://rust.godbolt.org/z/3bavf6rsT. Can you clarify what's going on there.

@rfcbot fcp merge

**Comment by compiler-errors** [comment]
@lcnr: Sorry. I thought that PR had caused the regression, but turns out that I had " fixed" it by making it `eq` then subsequently used `sup` because of regressions in `.` method dispatch: 8995c2c4a20197358605c3dad609e39f90a75ce8.

It was `sup` all along. So the state of the world before and after #122317 was the same, and the state of the world after this PR is objectively better.

**Comment by workingjubilee** [comment]
It seems that this started compiling in 1.0, stopped compiling in 1.54, started compiling again in 1.63, and will stop compiling again in 1.83:
https://rust.godbolt.org/z/4bxY7Gr4j

**Pull request merged**